### PR TITLE
Migrate Ballerina CLI to Golang

### LINKED ISSUE #42188: [Improvement]: Boost the Ballerina CLI performance with Golang
### Description

The Ballerina CLI is currently built as Bash and batch scripts. This approach lacks cross-platform compatibility, limiting the UX across diverse operating systems. We can rewrite certain parts of the CLI using Golang,  a modern and robust programming language known for its efficiency and cross-platform capabilities. The goal is to make the CLI cross-platform compatible making it easy to maintain and most importantly improve the performance and the user-friendliness of the Ballerina CLI.

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Other Area

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by Pabadhi**
Fixes #42188 

Create usage template
>Other
Version can be also handled using cobra.

## Check List 
- [ ] Create sample toml files for command generation
- [ ] Implement code to generate Cobra CLI commands
- [ ] Implement code to add tool commands dynamically
- [ ] Update build.gradle files
  - [ ] Implement Custom help command and custom help function
  - [ ] Implement Custom Help templates

### CODE REVIEW

**Comment by xlight05** [comment]
Let's rename the PR title to something like `Migrate Ballerina CLI to GO Cobra`.  Performance boost is an added bonus =)

File: cli/ballerina-go-cli/bal/cmd/help.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package cmd
```
Comment (xlight05) [inline cli/ballerina-go-cli/bal/cmd/help.go:1]: Add licence headers to all the applicable  files.

File: cli/ballerina-go-cli/executables/pkg/go.gradle [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    workingDir "$rootDir/cli/ballerina-go-cli/build"
+    commandLine './executable'
+}
+
+tasks.register('go_build_commands', Exec) {
+    group 'go'
+    description 'Compiles the packages named by the import paths, along with their dependencies. according to current code running architecture'
+    dependsOn updateGoScript
+    commandLine 'env', 'CGO_ENABLED=0', 'go', 'build', '-o',"$rootDir/cli/ballerina-go-cli/build/executable", "$rootDir/cli/ballerina-go-cli/executables/main.go"
+}
```
Comment (xlight05) [inline cli/ballerina-go-cli/executables/pkg/go.gradle:114]: Add a new line to EOF. check all the files

**Review by xlight05 — COMMENTED** [review]
cli/ballerina-go-cli/bal/LICENSE file seems empty. do we need it?

**Comment by xlight05** [comment]
Do we have an architecture diagram for this? It'll be great if we can have something for readability

File: cli/ballerina-cli/src/main/resources/command-info/asyncapi.toml [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+name = "input"
+shorthand = "i"
+default_val = ""
+usage = "This is a mandatory input. The listener will be generated according to the given AsyncAPI contract."
+
+[[base_command.flag]]
+name = "output"
+shorthand = "o"
+default_val = ""
+usage = "Location of the generated Ballerina source code."
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-cli/src/main/resources/command-info/asyncapi.toml:23]: Shall we add new line at the end of the file?

File: cli/ballerina-cli/src/main/resources/command-info/home.toml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[base_command]
+name = "home"
+short = "Display ballerina home"
+
+[help]
+    [help.base]
+    long = "long description of home"
+    examples="this is an example for home"
+    synopsis="bal home"
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-cli/src/main/resources/command-info/home.toml:9]: Add a new line at the end. Check other places as well.

File: cli/ballerina-go-cli/bal/main.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+/*
+Copyright © 2024 NAME HERE <EMAIL ADDRESS>
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/bal/main.go:2]: improper license

File: cli/ballerina-go-cli/bal/pkg/generate/generate.go [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+	viper.SetConfigType("json")
+	if err := viper.ReadInConfig(); err != nil {
+		log.Fatalln("Error reading config file:", err)
+	}
+	//Create Base command
+	toolID := viper.GetString("tool_id")
+	toolShort := viper.GetString("short")
+	toolLong := viper.GetString("help.base.long")
+	toolExamples := viper.GetString("help.base.examples")
+
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/bal/pkg/generate/generate.go:148]:
```suggestion
```

File: cli/ballerina-go-cli/bal/pkg/utils/utils.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+					javaHome = strings.TrimSpace(string(CmdOutput))
+				}
+			}
+		}
+
+	}
+
+	if javaCmd == "" {
+		if javaHome != "" {
+			javaCmdPath := filepath.Join(javaHome, "jre", "sh", "java")
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/bal/pkg/utils/utils.go:149]: shall we use a constant for `java`? Use constants for other possible places as well.

File: cli/ballerina-go-cli/bal/pkg/utils/utils.go [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+		}
+		balScriptPath = link
+	}
+	scriptPathDir := filepath.Dir(balScriptPath)
+	return scriptPathDir
+}
+
+func Setup() ([]string, string) {
+	ballerinaVersion := "@VERSION@"
+
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/bal/pkg/utils/utils.go:193]:
```suggestion
```

File: cli/ballerina-go-cli/bal/pkg/utils/utils.go [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+	scriptPathDir := GetBalScriptDir()
+
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/bal/pkg/utils/utils.go:195]:
```suggestion
```

File: cli/ballerina-go-cli/bal/pkg/utils/utils.go [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+	return cmdLineArgs, javaCmd
+}
+
+func ExecuteBallerinaCommand(javaCmd string, cmdLineArgs []string) error {
+	cmdArgs := append(cmdLineArgs, "io.ballerina.cli.launcher.Main")
+	cmdArgs = append(cmdArgs, os.Args[1:]...)
+	err := ExecuteCommand(javaCmd, cmdArgs)
+	return err
+
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/bal/pkg/utils/utils.go:248]:
```suggestion
```

File: cli/ballerina-go-cli/executables/pkg/go.gradle [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+tasks.register('go_windows_build_arm64', Exec) {
+    group 'go'
+    description 'Compiles the packages named by the import paths, along with their dependencies. ARM64 Architecture'
+    // Provide the path to the main Go file directly in the commandLine configuration
+    commandLine 'env', 'GOOS=windows', 'GOARCH=arm64','CGO_ENABLED=0', 'go', 'build','-o',"$rootDir/distribution/zip/jballerina/bin/bal_windows_arm64", "$rootDir/cli/ballerina-go-cli/build/bal/main.go"
+}
+
+tasks.register('go_build_default', Exec) {
```
Comment (Thevakumar-Luheerathan) [inline cli/ballerina-go-cli/executables/pkg/go.gradle:98]: shall we use `enum` for these tasks?

File: cli/ballerina-cli/src/main/resources/command-info/add.toml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+[base_command]
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/resources/command-info/add.toml:1]: missing license header

File: cli/ballerina-go-cli/bal/pkg/utils/utils.go [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+		if _, err := os.Stat(javaExePath); err != nil {
+			javaCmd = javaExePath
+		}
+	} else {
+		javaExePath := filepath.Join(javaHome, "bin", "java")
+		if _, err := os.Stat(javaExePath); err != nil {
+			javaCmd = javaExePath
+		}
+	}
+	//Set javaHome to installed java version in the user's device
```
Comment (azinneera) [inline cli/ballerina-go-cli/bal/pkg/utils/utils.go:127]:
```suggestion
	// Set javaHome to installed java version in the user's device
```

File: cli/ballerina-cli/src/main/resources/command-info/add.toml [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+name = "add"
+short = "short description of add"
```
Comment (gayaldassanayake) [inline cli/ballerina-cli/src/main/resources/command-info/add.toml:3]: Why don't we have the actual short description here?

Comment (Pabadhi) [reply cli/ballerina-cli/src/main/resources/command-info/add.toml:3]: I will add actual descriptions for all toml file. Here I added sample descriptions for testing.

**Pull request merged**